src/readsbml.jl

```diff
@@ -96,6 +96,39 @@ function readSBML(fn::String)::SBML.Model
             throw(AssertionError("SBML document contains no model"))
         end
 
+        for (converter, kwargs) in conversion_options
```

Comment:
extending the conversions this way requires you to extend SBML (thus basically making it non-extensible from the user side). You can pass in a normal conversion function to make this much more maintainable. (OOP folks would call it "callback" or "functor" or something)

Reply:
I've never worked with callbacks. Are you thinking of sth like this?

```Julia
# File readsbml.jl
function readSBML(fn::String;libsbml_converter::Maybe{Function}=nothing)::SBML.Model
    doc = ccall(sbml(:readSBML), VPtr, (Cstring,), fn)
    try
        ...

        # @Mirek: here happens the new conversion
        if !isnothing(libsbml_converter)
            libsbml_converter()
        end

        ...

        return extractModel(model)
    finally
        ccall(sbml(:SBMLDocument_free), Nothing, (VPtr,), doc)
    end
end


# File reactionsystem.jl
function libsbml_converter(conversion_options::Dict)

    for (converter, kwargs) in conversion_options
        if converter == "setLevelAndVersion"
            success = ccall(sbml(:SBMLDocument_setLevelAndVersion), Cint, (VPtr,Cint,Cint), doc, kwargs["level"], kwargs["version"])
            continue
        end
        props = ccall(sbml(:ConversionProperties_create), VPtr, ())
        option = ccall(sbml(:ConversionOption_create), VPtr, (Cstring,), converter)
        ccall(sbml(:ConversionProperties_addOption), Cvoid, (VPtr, VPtr), props, option)
        if !(kwargs==nothing)
            for (k, v) in kwargs
                option = ccall(sbml(:ConversionOption_create), VPtr, (Cstring,), k)
                ccall(sbml(:ConversionOption_setValue), Cvoid, (VPtr, Cstring), option, v)
                ccall(sbml(:ConversionProperties_addOption), Cvoid, (VPtr, VPtr), props, option)
            end
        success = ccall(sbml(:SBMLDocument_convert), Cint, (VPtr,VPtr), doc, props)
        end
    end
end

mtk_default_converter = function ()
    mtk_default_conversion_options = Dict("promoteLocalParameters" => nothing,
                                          "expandFunctionDefinitions" => nothing,
                                          "expandInitialAssignments" => nothing,
                                          "setLevelAndVersion" => Dict("level" => 3,
                                                                       "version" => 1)
                                         )
    libsbml_converter(mtk_default_conversion_options)
end

""" ReactionSystem constructor """
function ModelingToolkit.ReactionSystem(sbmlfile::String; libsbml_converter=mtk_default_converter, kwargs...)
    model = readSBML(sbmlfile;libsbml_converter=mtk_default_converter)
    ReactionSystem(model; kwargs...)
end
```

Reply:
@exaexa : I tried to merge master into my my pl/mk-species-units branch and this topic becomes relevant again. Should I handle the conversion like that now?
```Julia
# File readsbml.jl
function readSBML(fn::String, sbml_conversion = document -> nothing)::SBML.Model
    doc = ccall(sbml(:readSBML), VPtr, (Cstring,), fn)
    try
        ...

        sbml_conversion(doc)

        ...

        return extractModel(model)
    finally
        ccall(sbml(:SBMLDocument_free), Nothing, (VPtr,), doc)
    end
end


# File reactionsystem.jl
function libsbml_converter(doc; conversion_options::Dict=Dict("promoteLocalParameters" => nothing, "expandFunctionDefinitions" => nothing, "expandInitialAssignments" => nothing, "setLevelAndVersion" => Dict("level" => 3, "version" => 1)))

    for (converter, kwargs) in conversion_options
        if converter == "setLevelAndVersion"
            success = ccall(sbml(:SBMLDocument_setLevelAndVersion), Cint, (VPtr,Cint,Cint), doc, kwargs["level"], kwargs["version"])
            continue
        end
        props = ccall(sbml(:ConversionProperties_create), VPtr, ())
        option = ccall(sbml(:ConversionOption_create), VPtr, (Cstring,), converter)
        ccall(sbml(:ConversionProperties_addOption), Cvoid, (VPtr, VPtr), props, option)
        if !(kwargs==nothing)
            for (k, v) in kwargs
                option = ccall(sbml(:ConversionOption_create), VPtr, (Cstring,), k)
                ccall(sbml(:ConversionOption_setValue), Cvoid, (VPtr, Cstring), option, v)
                ccall(sbml(:ConversionProperties_addOption), Cvoid, (VPtr, VPtr), props, option)
            end
        success = ccall(sbml(:SBMLDocument_convert), Cint, (VPtr,VPtr), doc, props)
        end
    end
end

""" ReactionSystem constructor """
function ModelingToolkit.ReactionSystem(sbmlfile::String; libsbml_converter=libsbml_converter, kwargs...)
    model = readSBML(sbmlfile;libsbml_converter=mtk_default_converter)
    ReactionSystem(model; kwargs...)
end
```

Reply:
Lets just create one converter:
```Julia
function libsbml_converter!(doc, converter::String; kwargs...)
    ...
    return doc
end

```
So you could then do
```Julia
readSBML("filename.xml", doc->begin
  libsbml_converter!(doc,"setLevelAndVersion"; level=3, version=1)
  libsbml_converter!(doc,"promoteLocalParameters")
end)
```

Reply:
That works now perfectly already:
```
readSBML("filename.xml", doc -> begin
  convert_level_and_version(3,1)(doc)
  libsbml_convert(Dict("promoteLocalParameters" => Dict{String,String}())
    )(doc)
end)
```
The (doc) parameters are curried to allow pointfree specification for single converters, but it works the same as before.
I could simplify the API of libsbml_convert though, the dictionary there looks more confusing then helpful.